src/main/java/org/tugraz/sysds/runtime/lineage/LineageCache.java

```diff
@@ -1,60 +1,307 @@
+/*
+ * Copyright 2019 Graz University of Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.tugraz.sysds.runtime.lineage;
 
 import org.tugraz.sysds.api.DMLScript;
+import org.tugraz.sysds.hops.OptimizerUtils;
+import org.tugraz.sysds.hops.cost.CostEstimatorStaticRuntime;
+import org.tugraz.sysds.lops.MMTSJ.MMTSJType;
+import org.tugraz.sysds.runtime.DMLRuntimeException;
 import org.tugraz.sysds.runtime.controlprogram.caching.MatrixObject;
 import org.tugraz.sysds.runtime.controlprogram.context.ExecutionContext;
+import org.tugraz.sysds.runtime.instructions.CPInstructionParser;
 import org.tugraz.sysds.runtime.instructions.Instruction;
+import org.tugraz.sysds.runtime.instructions.cp.CPInstruction.CPType;
 import org.tugraz.sysds.runtime.instructions.cp.ComputationCPInstruction;
+import org.tugraz.sysds.runtime.instructions.cp.MMTSJCPInstruction;
 import org.tugraz.sysds.runtime.matrix.data.MatrixBlock;
+import org.tugraz.sysds.runtime.util.LocalFileUtils;
 
+import java.io.IOException;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
 
 public class LineageCache {
-	private static final Map<LineageItem, MatrixBlock> _cache = new HashMap<>();
-	
+	private static final Map<LineageItem, Entry> _cache = new HashMap<>();
+	private static final Map<LineageItem, String> _spillList = new HashMap<>();
+	private static final HashSet<LineageItem> _removelist = new HashSet<>();
+	private static final long CACHELIMIT= (long)512*1024*1024; // 500MB
+	private static String outdir = null;
+	private static long _cachesize = 0;
+	private static Entry _head = null;
+	private static Entry _end = null;
+
+	//--------------------- CACHE LOGIC METHODS ----------------------
+
 	public static void put(Instruction inst, ExecutionContext ec) {
 		if (!DMLScript.LINEAGE_REUSE)
 			return;
-		
-		if( inst instanceof ComputationCPInstruction
-			&&((ComputationCPInstruction) inst).output.getDataType().isMatrix() ) {
-			
-			for (LineageItem item : ((LineageTraceable) inst).getLineageItems()) {
+
+		if (inst instanceof ComputationCPInstruction &&
+			inst.getOpcode().equalsIgnoreCase("tsmm")) {
+			LineageItem[] items = ((LineageTraceable) inst).getLineageItems();
+			for (LineageItem item : items) {
 				MatrixObject mo = ec.getMatrixObject(((ComputationCPInstruction) inst).output);
-				LineageCache._cache.put(item, mo.acquireReadAndRelease());
+				LineageCache.put(inst, item, mo.acquireRead());
```

Comment:
Please use the shorter `mo.acquireReadAndRelease()`

Reply:
Also please add a comment that it's safe to pin the object here, because it comes from a CPInstruction and hence is guaranteed to fit into the driver.